DependencyInjection/Configuration.php

```diff
@@ -40,26 +43,91 @@ public function getConfigTreeBuilder() : TreeBuilder
         $organizeMigrationModes = $this->getOrganizeMigrationsModes();
 
         $rootNode
+            ->fixXmlConfig('migration', 'migrations')
+            ->fixXmlConfig('migrations_path', 'migrations_paths')
             ->children()
-                ->scalarNode('dir_name')->defaultValue('%kernel.root_dir%/DoctrineMigrations')->cannotBeEmpty()->end()
-                ->scalarNode('namespace')->defaultValue('Application\Migrations')->cannotBeEmpty()->end()
-                ->scalarNode('table_name')->defaultValue('migration_versions')->cannotBeEmpty()->end()
-                ->scalarNode('column_name')->defaultValue('version')->end()
-                ->scalarNode('column_length')->defaultValue(14)->end()
-                ->scalarNode('executed_at_column_name')->defaultValue('executed_at')->end()
-                ->scalarNode('all_or_nothing')->defaultValue(false)->end()
                 ->scalarNode('name')->defaultValue('Application Migrations')->end()
-                ->scalarNode('custom_template')->defaultValue(null)->end()
-                ->scalarNode('check_database_platform')->defaultValue(true)->end()
-                ->scalarNode('organize_migrations')->defaultValue(false)
+
+                ->arrayNode('migrations_paths')
+                    ->info('A list of namespace/path pairs where to look for migrations.')
+                    ->isRequired()
+                    ->requiresAtLeastOneElement()
+                    ->useAttributeAsKey('namespace')
+                    ->prototype('scalar')->end()
+                 ->end()
+
+                ->arrayNode('services')
+                    ->info('A set of services to pass to the underlying doctrine/migrations library, allowing to change its behaviour.')
+                    ->useAttributeAsKey('service')
+                    ->defaultValue([])
+                    ->validate()
+                        ->ifTrue(static function ($v) {
+                            return count(array_filter(array_keys($v), static function (string $doctrineService) : bool {
+                                return strpos($doctrineService, 'Doctrine\Migrations\\') !==0;
+                            }));
+                        })
+                        ->thenInvalid('Valid services for the DoctrineMigrationsBundle must be in the "Doctrine\Migrations" namespace.')
```

Comment:
Should we allow users to use custom services for certain things? I'm not sure what good it does to expose this in the configuration if users can only use our own services - we might as well configure them in the extension. Can you outline the use-case you have for this?

Reply:
My usecase is the ability to change the migrations factory.  being able to inject extra dependencies in the migration class as  the symfony container or similar.